### ISSUE TITLE
--event-template is poorly documented and tested

### ISSUE THREAD

**Issue by CooperBarrett** [tobixen/calendar-cli#45]
Sorry I use a translator (Deepl.com)
----
hello,
First of all, thanks for calendar-cli, it's really great.
I would like to know if there would be (or if there is already) the possibility to add a description and the place to the agenda?
I use descriptions a lot in calendar events and that would be really useful for me.
Thanks in advance for your help.

**Comment by tobixen** [comment]
Use `--set-description` - this should work:

    calendar-cli calendar add '2018-04-17 16:00:00+2h' 'Project meeting' --set-description 'we will discuss various details and agree on some other details related to the project' --set-location 'Meeting room 3, floor 5, Krugen office'

In the agenda mode it should be possible to use `--event-template` to display the description:

    calendar-cli calendar agenda --event-template '{dtstart} {summary} {description}'

however, this is poorly documented AND it currently seems to break, so I will have to look more into it.  Leaving this issue as open until it's fixed.

**Comment by CooperBarrett** [comment]
Sorry, I'm reopening this topic for a quick question.

Is it possible to add a "full day" event? I am not talking about a 24-hour event, I do not know if I am clear...
**Event Full Day** : 
![](https://framapic.org/xmO8UndH3ZqG/AaaEvK4RoQLR)
**Event 24h (23h59)** : 
![](https://framapic.org/ExxnCxM5QDu6/efajtWd1rfrq)

**Comment by CooperBarrett** [comment]
I'm going to open another issue out because it's a different question

**Comment by tobixen** [comment]
This bug report should be left open

**Comment by CooperBarrett** [comment]
I'm currently trying to use --event-template but there are indeed problems.
This does not work with 
 - {description}
 - {location}
Is there a way to display them or is it still being tested ?

**Comment by tobixen** [comment]
I just made a commit to the master branch that will fix description and location, in master branch.

Prior to that commit, it may be possible to use {instance.description.value} ... like this:

    $ calendar-cli calendar agenda --event-template '{dstart} {instance.description.value}'
    2018-12-20 07:20 (Thu) Fly med Norwegian, reservasjon KQPANL. Fra Oslo-Gardermoen til Troms&#248; 20. des 2018 08:20, DY370

But then one will get into two issues - for one thing, description is often a multi-line-thingy.  Trying this with my work calendar, it appears our weekly department meeting has several hundred lines of meaningless description as some ten-twenty lines have been added automatically was added every time the invitee-list, meeting-room or other details got updated.  The other issue is that eventually it will break once it hits an event that has no description.

I think both calendar_agenda and todo_list should be refactored, I have some thoughts on how to do this properly:

* revert the commit
* fortify the test code covering --event-template and create a test covering --todo-template
* fix a function that will take a template and fetch all "simple" identifiers used (maybe using some regexp like \{(\w+)[\:\!\}]
* for every identifier attr that is used ...
    * if attr in event: continue
    * if hasattr(event['instance'], attr): event[attr] = getattr(event['instance'], attr).value
    * eventually, more tweaking in case it's a datetime, a list or something else than a string

**Comment by CooperBarrett** [comment]
Sorry for the response time, I was on the weekend.
Thank you for the answer, indeed `{instance.description.value}` only works if there is a description of a registrant.
I only have short descriptions so no problem.

I also have this error if an accent is in the description:
```
Traceback (most recent call last):
  File "/usr/local/bin/bin/calendar-cli.py", line 4, in <module>
    __import__('pkg_resources').run_script('calendar-cli==0.0.0','calendar-cli.py')
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 739, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 1501, in run_script
    exec(script_code, namespace, namespace)
  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.0.0.0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 870, in <module>

  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.0.0.0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 867, in main

  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.0.0.0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 474, in calendar_agenda

UnicodeEncodeError:'ascii' codec can't encode character u'\xe8' in position 4: ordinal not in range(128)
```

These problems are also valid for `{instance.location.value}`.
It's more embarrassing because for `description`, I can add a meaningless description while for `location` I can't put a place that doesn't match anything.

In short, thank you for what you are doing, I hope you will be able to solve these problems.

Thank you in advance

Translated with www.DeepL.com/Translator

**Comment by tobixen** [comment]
If you check out the latest master branch, you should be able to do:

$ calendar-cli calendar agenda --event-template '{dtstart} {summary} {description} {location}'

**Comment by CooperBarrett** [comment]
Thank you very much !

It works almost perfectly. I have this last concern about accents (we have a lot of them in French...) in `{description}` or `{location}`. 
On the other hand, no problem if there are any in `{summary}`.

In any case, it's a big step forward for me, thank you very much for your reactivity.

**Comment by CooperBarrett** [comment]
Ah... Strange.

I have just done some tests again and I confirm my problem with the accents.
I have this error review:
```
Traceback (most recent call last):
  File "/usr/local/bin/bin/calendar-cli.py", line 4, in <module>
    __import__('pkg_resources').run_script('calendar-cli==0.11.0.dev0','calendar-cli.py')
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 739, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 1501, in run_script
    exec(script_code, namespace, namespace)
  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.11.0.dev0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 904, in <module>

  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.11.0.dev0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 901, in main

  File "/usr/local/lib/python2.7/dist-packages/calendar_cli-0.11.0.dev0-py2.7.egg/EGG-INFO/scripts/calendar-cli.py", line 505, in calendar_agenda

UnicodeEncodeError:'ascii' codec can't encode character u'\xeb' in position 7: ordinal not in range(128)
```
In this example, the problematic word is **Noël** and the complete description is: `F0 = Noël`

Could this be from my locals?

**Comment by tobixen** [comment]
Came to think, I only tested with python3, not python2

**Comment by tobixen** [comment]
nevermind, calendar-cli is hardcoded to use python2, so that can't be the problem.

**Comment by CooperBarrett** [comment]
ok, for information, i use 
**Python 2.7.13**

**Comment by tobixen** [comment]
try again, found an obvious bug :-)

**Comment by CooperBarrett** [comment]
You're the best!
It works perfectly, thank you very much !

**Comment by tobixen** [comment]
So what remains here is to document the event-templates and ensure todo-templates works as good as event-templates.  It could be a simple job, just cut'n'paste and edit the work I've done with events slightly.  However, I really hate code duplication, so I think it's important to do some refactoring.

**Comment by tobixen** [comment]
Refactoring of templates should be done as part of task #12, documentation in task #65.  I'll close this issue now.
